fastestimator/backend/focal_loss.py

```diff
@@ -0,0 +1,145 @@
+            input_data: Floating point value.
+
+        Returns:
+            Converted input_data values.
+
+        Raises:
+            AssertionError: If `y_true` are unacceptable data types. if data type is other than np.array, tensor.Tensor, tf.Tensor.
+    """
+    if tf.is_tensor(y_true):
+        return tf.cast(input_data, y_true.dtype)
+    elif isinstance(y_true, torch.Tensor):
+        return torch.tensor(input_data).type(y_true.dtype)
+    else:
+        raise ValueError("Unsupported tensor type.")
+
+
+def focal_loss(y_true: Tensor, y_pred: Tensor, gamma: float = 2.0, alpha: float = 0.25,  from_logits: bool = False, reduction: str = "mean") -> Tensor:
+    """
+    Calculate the focal loss between two tensors.
+    Original implementation from https://github.com/facebookresearch/fvcore/blob/master/fvcore/nn/focal_loss.py .
+    Loss used in RetinaNet for dense detection: https://arxiv.org/abs/1708.02002.
+
+    This method can be used with TensorFlow tensors:
+    ```python
+    true = tf.constant([[1], [1], [1], [0], [0], [0]])
+    pred = tf.constant([[0.97], [0.91], [0.73], [0.27], [0.09], [0.03]])
+    b = fe.backend.focal_loss(y_pred=pred, y_true=true, gamma=None, alpha=None) #0.1464
+    b = fe.backend.focal_loss(y_pred=pred, y_true=true, gamma=2.0, alpha=0.25) #0.00395
+    ```
+
+    This method can be used with PyTorch tensors:
+    ```python
+    true = torch.tensor([[1], [1], [1], [0], [0], [0]])
+    pred = torch.tensor([[0.97], [0.91], [0.73], [0.27], [0.09], [0.03]])
+    b = fe.backend.focal_loss(y_pred=pred, y_true=true, gamma=None, alpha=None) #0.1464
+    b = fe.backend.focal_loss(y_pred=pred, y_true=true, gamma=2.0, alpha=0.25) #0.004
+    ```
+
+    Args:
+        y_true: Ground truth class labels which should take values of 1 or -1.
+        y_pred: Prediction score for each class, with a shape like y_true. dtype: float32 or float16.
+        alpha: Weighting factor in range (0,1) to balance
+                positive vs negative examples or -1 for ignore. Default = 0.25
+        gamma: Exponent of the modulating factor (1 - p_t) to
+               balance easy vs hard examples.
+        from_logits: Whether y_pred is logits (without softmax).
+        reduction: 'none' | 'mean' | 'sum'
+                 'none': No reduction will be applied to the output.
+                 'mean': The output will be averaged.
+                 'sum': The output will be summed.
+    Returns:
+        The Focal loss between `y_true` and `y_pred`
+
+    Raises:
+        ValueError: If `y_pred` or 'y_true' is an unacceptable data type.
+    """
+    if gamma and gamma < 0:
+        raise ValueError(
+            "Value of gamma should be greater than or equal to zero.")
+
+    if alpha:
+        alpha = cast(alpha, y_pred)
+
+    if gamma:
+        gamma = cast(gamma, y_pred)
+
+    if isinstance(y_true, tf.Tensor):
```

Comment:
I think this needs to be tf.is_tensor() rather than isinstance() in order to work with variables / multi -gpu